File: C++/Problems/Problems/problem_826.cpp
#include "problem_826.h"

bool problem_826::test() {
    std::vector difficulty {2, 4, 6, 8, 10};
    std::vector profit {10, 20, 30, 40, 50};
    std::vector worker { 4, 5, 6, 7 };

    int expected = 100;

    auto result = maxProfitAssignment(difficulty, profit, worker);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_826::maxProfitAssignment(std::vector<int> &difficulty, std::vector<int> &profit, std::vector<int> &worker) {
    int n = difficulty.size();
    int m = worker.size();
    int maxSkill = 0;
    for (int i = 0; i < m; ++i)
        maxSkill = std::max(maxSkill, worker[i]);
    std::vector bestJobs(maxSkill + 1, 0);
    for (int i = 0; i < n; ++i)
    {
        if (difficulty[i] <= maxSkill)
            bestJobs[difficulty[i]] = std::max(bestJobs[difficulty[i]], profit[i]);
    }
    int curr = 0;
    for (int i = 0; i <= maxSkill; ++i)
    {
        curr = std::max(curr, bestJobs[i]);
        bestJobs[i] = curr;
    }
    int result = 0;
    for (int i = 0; i < m; ++i)
        result += bestJobs[worker[i]];

    return result;
}

File: Python/Problems/Problems/problem826.py
import heapq
import math
import sys

import numpy as np

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem

class Problem826(Problem):
    def test(self) -> bool:
        difficulty = [2, 4, 6, 8, 10]
        profit = [10, 20, 30, 40, 50]
        worker = [4, 5, 6, 7]

        expected = 100

        result = self.maxProfitAssignment(difficulty, profit, worker)

        print(result)

        return result == expected

    def maxProfitAssignment(self, difficulty: List[int], profit: List[int], worker: List[int]) -> int:
        n = len(difficulty)
        m = len(worker)
        maxSkill = 0
        for i in range(m):
            maxSkill = max(maxSkill, worker[i])
        bestJobs = [0] * (maxSkill + 1)
        for i in range(n):
            if difficulty[i] <= maxSkill:
                bestJobs[difficulty[i]] = max(bestJobs[difficulty[i]], profit[i])

        curr = 0
        for i in range(maxSkill + 1):
            curr = max(curr, bestJobs[i])
            bestJobs[i] = curr
        result = 0
        for i in range(m):
            result += bestJobs[worker[i]]

        return result